test_volume.py

```diff
@@ -178,3 +179,198 @@ def test_maximum_number_of_volumes(server, create_volume):
     assert error.value.response.json()['detail'] == (
         "Due to internal limitations, it is currently not possible "
         "to attach more than 128 volumes.")
+
+
+def test_snapshot_volume_attached(server, volume):
+    """ Attached volumes can be snapshotted and reverted.
+
+    It is possible to create a snapshot of a volume which is currently attached
+    and to revert the volume back to this state.
+
+    Snapshots of volumes taken while they are attached and mounted are crash
+    consistent. Some in-flight data might not be in the snapshot, but volume
+    can always be recovered to a consistent state.
+    """
+
+    # Attach and format volume to server
+    volume.attach(server)
+    time.sleep(5)
+    server.assert_run('sudo mkfs.ext4 /dev/sdb')
+    server.assert_run('sudo mount /dev/sdb /mnt')
+
+    # Create two files. The first is synced to disk with fsync, the second
+    # is not synced. Data might still be in-flight.
+    server.assert_run(
+        'sudo dd if=/dev/zero of=/mnt/synced count=1 bs=1M conv=fsync')
+    server.assert_run('sudo dd if=/dev/zero of=/mnt/not-synced count=1 bs=1M')
+
+    # Create snapshot
+    snapshot = volume.snapshot('snap')
+
+    # Write test file to volume
+    server.assert_run('sudo touch /mnt/after-snapshot')
+
+    # Try reverting while attached (should fail)
+    with pytest.raises(HTTPError) as error:
+        volume.revert(snapshot)
+
+    # Assert a HTTP 400 BadRequest response with a specific error message
+    assert error.value.response.status_code == 400
+    assert error.value.response.json()['detail'] == (
+        'Cannot revert non-root volumes while they are attached to a server.'
+    )
+
+    # Detach volume
+    server.assert_run('sudo umount /mnt')
+    volume.detach()
+    time.sleep(5)
+
+    # Revert volume to snapshot
+    volume.revert(snapshot)
+
+    # Reattach and mount the volume
+    volume.attach(server)
+    time.sleep(5)
+    server.assert_run('sudo mount /dev/sdb /mnt')
+
+    # Verify the test files are in the correct state
+    assert server.file_path_exists('/mnt/synced')
+    assert not server.file_path_exists('/mnt/after-snapshot')
+
+    # "Warn" if the file "not-synced" exists. This is not a failure because
```

Comment:
I'm not sure what the use of this is. I don't think this is a warning, as that is a legitimate state of affairs. I would probably just drop this.

Reply:
As acceptance tests are also meant to show how a feature works, I think this nicely shows that you have to sync data to ensure it's in the snapshot and that this is not only a theoretical concern. So I would prefer to keep it in some form.

I agree that `warn` is suboptimal. Is there another way to log this?

Reply:
My point is mainly, that whatever we do, if we do not assert, then this is just going to end up in an output nobody reads. I guess as a form of comment it's okay.